File: Leetcode 567 - permutation-in-string/567.permutation-in-string.cpp
/*
 * @lc app=leetcode id=567 lang=cpp
 *
 * [567] Permutation in String
 */

// @lc code=start
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
  bool checkInclusion(string s1, string s2) {
    // unordered_map<char, int> s1Map;
    // for (char c : s1) {
    //   s1Map[c]++;
    // }

    vector<int> s1LetterCount(26, 0);
    for (int i = 0; i < s1.size(); i++) {
      s1LetterCount[s1[i] - 'a']++;
    }

    int left = 0, right = s1.size() - 1;

    while (right < s2.size()) {
      //   unordered_map<char, int> s1MapCopy = s1Map;
      //   for (int i = left; i <= right; i++) {
      //     if (s1MapCopy.find(s2[i]) != s1MapCopy.end()) {
      //       s1MapCopy[s2[i]]--;
      //     }
      //     if (s1MapCopy[s2[i]] == 0) {
      //       s1MapCopy.erase(s2[i]);
      //     }
      //   }

      //   if (s1MapCopy.empty()) {
      //     return true;
      //   }

      vector<int> s2LetterCount(26, 0);

      for (int i = left; i <= right; i++) {
        s2LetterCount[s2[i] - 'a']++;
      }

      if (s1LetterCount == s2LetterCount)
        return true;

      left++;
      right++;
    }

    return false;
  }
};

int main() {
  Solution sol;

  cout << boolalpha << endl;
  cout << sol.checkInclusion("ab", "eidbaooo") << endl; // true
  cout << sol.checkInclusion("ab", "eidboaoo") << endl; // false
}
// @lc code=end

File: Leetcode 567 - permutation-in-string/567.permutation-in-string.java
/*
 * @lc app=leetcode id=567 lang=java
 *
 * [567] Permutation in String
 */

// @lc code=start

class Solution {

  public boolean checkInclusion(String s1, String s2) {
    if (s1.length() > s2.length()) return false;

    int[] letters = new int[26];

    // Count the letters in s1
    for (int i = 0; i < s1.length(); i++) {
      letters[s1.charAt(i) - 'a']++;
    }

    int left = 0, right = 0;

    // Slide the window
    while (right < s2.length()) {
      // Remove the letter from the counter
      letters[s2.charAt(right) - 'a']--;

      // If the window size is the same as s1, check if all letters are 0
      if (right - left + 1 == s1.length()) {
        boolean allZero = true;
        for (int count : letters) {
          if (count != 0) {
            allZero = false;
            break;
          }
        }

        // If all letters are 0, return true
        if (allZero) {
          return true;
        }

        // Add the letter back to the counter
        letters[s2.charAt(left) - 'a']++;
        left++;
      }

      right++;
    }

    return false;
  }

  public static void main(String[] args) {
    Solution solution = new Solution();

    System.out.println(solution.checkInclusion("ab", "eidbaooo")); // true
    System.out.println(solution.checkInclusion("ab", "eidboaoo")); // false
  }
}
// @lc code=end
